src/transformers/models/timesformer/modeling_timesformer.py

```diff
@@ -0,0 +1,845 @@
+# coding=utf-8
+# Copyright 2022 Meta and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch TimeSformer model."""
+
+
+import collections
+import math
+import warnings
+from typing import Optional, Tuple, Union
+
+import torch
+import torch.nn.functional
+import torch.utils.checkpoint
+from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+
+from ...activations import ACT2FN
+from ...modeling_outputs import BaseModelOutput, ImageClassifierOutput
+from ...modeling_utils import PreTrainedModel
+from ...utils import add_start_docstrings, add_start_docstrings_to_model_forward, logging, replace_return_docstrings
+from .configuration_timesformer import TimeSformerConfig
+
+
+logger = logging.get_logger(__name__)
+
+_CONFIG_FOR_DOC = "TimeSformerConfig"
+_CHECKPOINT_FOR_DOC = "facebook/timesformer"
+
+TIMESFORMER_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "fcakyon/timesformer-base-finetuned-k400",
+    # See all TimeSformer models at https://huggingface.co/models?filter=timesformer
+]
+
+
+# Adapted from https://github.com/facebookresearch/TimeSformer/blob/a5ef29a7b7264baff199a30b3306ac27de901133/timesformer/models/vit.py#L155
+class TimeSformerPatchEmbeddings(nn.Module):
+    """Image to Patch Embedding"""
+
+    def __init__(self, config):
+        super().__init__()
+
+        image_size = config.image_size
+        patch_size = config.patch_size
+
+        image_size = image_size if isinstance(image_size, collections.abc.Iterable) else (image_size, image_size)
+        patch_size = patch_size if isinstance(patch_size, collections.abc.Iterable) else (patch_size, patch_size)
+
+        num_patches = (image_size[1] // patch_size[1]) * (image_size[0] // patch_size[0])
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.num_patches = num_patches
+
+        self.projection = nn.Conv2d(config.num_channels, config.hidden_size, kernel_size=patch_size, stride=patch_size)
+
+    def forward(self, pixel_values):
+        batch_size, num_channels, num_frames, height, width = pixel_values.shape
+        pixel_values = pixel_values.permute(0, 2, 1, 3, 4).reshape(
+            batch_size * num_frames, num_channels, height, width
+        )
```

Comment:
```suggestion
        batch_size, num_frames, num_channels, height, width = pixel_values.shape
        pixel_values = pixel_values.reshape(batch_size * num_frames, num_channels, height, width)
```
Wouldn't this suffice? Also, no need to permute another time in the `TimeSformerEmbeddings` class.

Reply:
It doesnt work. The dimensions should be rearranged before the reshape call.

When I switch to your suggestion all tests fails with such errors:
```bash
________________________________________________________ TimesformerModelTest.test_retain_grad_hidden_states_attentions _________________________________________________________ 
[gw1] win32 -- Python 3.9.12 C:\Users\FCA\Miniconda3\envs\transformers\python.exe

self = <tests.models.timesformer.test_modeling_timesformer.TimesformerModelTest testMethod=test_retain_grad_hidden_states_attentions>

    def test_retain_grad_hidden_states_attentions(self):
        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
        config.output_hidden_states = True
        config.output_attentions = self.has_attentions

        # no need to test all models as different heads yield the same functionality
        model_class = self.all_model_classes[0]
        model = model_class(config)
        model.to(torch_device)

        inputs = self._prepare_for_class(inputs_dict, model_class)

>       outputs = model(**inputs)

tests\test_modeling_common.py:1194:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
src\transformers\models\timesformer\modeling_timesformer.py:632: in forward
    embedding_output = self.embeddings(pixel_values)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
src\transformers\models\timesformer\modeling_timesformer.py:109: in forward
    embeddings, num_frames, patch_width = self.patch_embeddings(pixel_values)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
src\transformers\models\timesformer\modeling_timesformer.py:71: in forward
    embeddings = self.projection(pixel_values)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\conv.py:463: in forward
    return self._conv_forward(input, self.weight, self.bias)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Conv2d(3, 32, kernel_size=(2, 2), stride=(2, 2))
input = tensor([[[[0.1734, 0.2185, 0.4545,  ..., 0.5088, 0.2298, 0.8207],
          [0.2557, 0.4301, 0.6091,  ..., 0.3722, 0.0...76, 0.4126, 0.2142,  ..., 0.8532, 0.4099, 0.5987],
          [0.7795, 0.4282, 0.8127,  ..., 0.5824, 0.1563, 0.4537]]]])
weight = Parameter containing:
tensor([[[[-5.4493e-03,  3.8472e-02],
          [ 4.3009e-02,  3.3034e-02]],

         [[ 1.8041...-03,  5.4914e-02]],

         [[-2.3999e-03,  7.3581e-03],
          [ 3.2404e-02,  1.0074e-02]]]], requires_grad=True)
bias = Parameter containing:
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0.], requires_grad=True)

    def _conv_forward(self, input: Tensor, weight: Tensor, bias: Optional[Tensor]):
        if self.padding_mode != 'zeros':
            return F.conv2d(F.pad(input, self._reversed_padding_repeated_twice, mode=self.padding_mode),
                            weight, bias, self.stride,
                            _pair(0), self.dilation, self.groups)
>       return F.conv2d(input, weight, bias, self.stride,
                        self.padding, self.dilation, self.groups)
E       RuntimeError: Given groups=1, weight of size [32, 3, 2, 2], expected input[39, 2, 10, 10] to have 3 channels, but got 2 channels instead

..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\conv.py:459: RuntimeError
```

```bash
__________________________________________________________________ TimesformerModelTest.test_attention_outputs __________________________________________________________________ 
[gw1] win32 -- Python 3.9.12 C:\Users\FCA\Miniconda3\envs\transformers\python.exe

self = <tests.models.timesformer.test_modeling_timesformer.TimesformerModelTest testMethod=test_attention_outputs>

    def test_attention_outputs(self):
        if not self.has_attentions:
            pass

        else:
            config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
            config.return_dict = True

            for model_class in self.all_model_classes:
                seq_len = self.model_tester.seq_length
                num_frames = self.model_tester.num_frames

                inputs_dict["output_attentions"] = True
                inputs_dict["output_hidden_states"] = False
                config.return_dict = True
                model = model_class(config)
                model.to(torch_device)
                model.eval()
                with torch.no_grad():
>                   outputs = model(**self._prepare_for_class(inputs_dict, model_class))

tests\models\timesformer\test_modeling_timesformer.py:231:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
src\transformers\models\timesformer\modeling_timesformer.py:632: in forward
    embedding_output = self.embeddings(pixel_values)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
src\transformers\models\timesformer\modeling_timesformer.py:109: in forward
    embeddings, num_frames, patch_width = self.patch_embeddings(pixel_values)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
src\transformers\models\timesformer\modeling_timesformer.py:71: in forward
    embeddings = self.projection(pixel_values)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\module.py:1190: in _call_impl
    return forward_call(*input, **kwargs)
..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\conv.py:463: in forward
    return self._conv_forward(input, self.weight, self.bias)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Conv2d(3, 32, kernel_size=(2, 2), stride=(2, 2))
input = tensor([[[[8.5111e-01, 3.3832e-01, 6.6561e-01,  ..., 9.5386e-02,
           7.3812e-01, 1.9491e-01],
          [6.0208...1, 1.1054e-01],
          [9.3212e-01, 4.9163e-01, 5.0713e-01,  ..., 4.3621e-01,
           3.9658e-01, 1.6134e-01]]]])
weight = Parameter containing:
tensor([[[[-1.3263e-03, -2.0188e-02],
          [ 4.4333e-03,  2.2196e-02]],

         [[ 1.6649...-02, -3.5418e-02]],

         [[ 4.6990e-02, -3.2687e-02],
          [-1.0032e-02, -3.3334e-03]]]], requires_grad=True)
bias = Parameter containing:
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0.], requires_grad=True)

    def _conv_forward(self, input: Tensor, weight: Tensor, bias: Optional[Tensor]):
        if self.padding_mode != 'zeros':
            return F.conv2d(F.pad(input, self._reversed_padding_repeated_twice, mode=self.padding_mode),
                            weight, bias, self.stride,
                            _pair(0), self.dilation, self.groups)
>       return F.conv2d(input, weight, bias, self.stride,
                        self.padding, self.dilation, self.groups)
E       RuntimeError: Given groups=1, weight of size [32, 3, 2, 2], expected input[39, 2, 10, 10] to have 3 channels, but got 2 channels instead

..\..\Miniconda3\envs\transformers\lib\site-packages\torch\nn\modules\conv.py:459: RuntimeError
```

Reply:
Thanks a lot for bringing this up. I dont know how I missed it. Removed he double permute now 👍